File: fixtures/scale/scala/large/src/cryptobench/ciphers/beaufort/BeaufortCipher.scala
package cryptobench.ciphers.beaufort

import cryptobench.core.Cipher
import cryptobench.util.Alphabet

/** Reciprocal variant: ciphertext is key letter minus plaintext letter. */
final class BeaufortCipher(key: BeaufortKey) extends Cipher {

  override def name: String = "beaufort"

  override def encrypt(plaintext: String): String = transform(Alphabet.clean(plaintext), forward = true)

  override def decrypt(ciphertext: String): String = transform(Alphabet.clean(ciphertext), forward = false)

  private def transform(text: String, forward: Boolean): String = {
    val sb = new StringBuilder
    for (i <- text.indices) {
      val x = Alphabet.indexOf(text.charAt(i))
      val k = Alphabet.indexOf(key.keyCharAt(i))
      sb.append(Alphabet.charAt(if (forward) k - x else k - x))
    }
    sb.toString
  }
}

File: fixtures/scale/dart/large/lib/ciphers/beaufort/beaufort_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'beaufort_key.dart';

/// Reciprocal variant: ciphertext is key letter minus plaintext letter.
class BeaufortCipher implements Cipher {
  final BeaufortKey key;

  BeaufortCipher(this.key);

  @override
  String name() => 'beaufort';

  @override
  String encrypt(String plaintext) => _transform(cleanAlphabet(plaintext), true);

  @override
  String decrypt(String ciphertext) => _transform(cleanAlphabet(ciphertext), false);

  String _transform(String text, bool forward) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final k = indexOf(key.keyCharAt(i));
      sb.write(charAt(forward ? k - x : k - x));
    }
    return sb.toString();
  }
}
